TST: fix tests for mixed int string index

natmokval opened the pull request:
- [x] closes #54072
Added in conftest's `indices_dict` an entry:
```
"mixed-int-string": Index([0, "a", 1, "b", 2, "c"])
```
Adding "mixed-int-string" causes some tests to fail, fixed these tests.

natmokval committed:
add to indices_dict entry mixed-int-string

natmokval committed:
fix tests in test_old_base.py, test_numpy_compat.py

natmokval committed:
add except TypeError to def nargsort, fix tests

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
fix tests

jbrockmendel commented on pandas/core/sorting.py:
```diff
@@ -436,7 +436,13 @@ def nargsort(
+    try:
+        indexer = non_nan_idx[non_nans.argsort(kind=kind)]
+    except TypeError as err:
+        msg = "'<' not supported between "
```
this looks like the message just cuts off?  what is the message in err?

natmokval replied:
Thank you for the comments.
Yes, I cut the message off because in tests I had both: "'<' not supported between instances of 'int' and 'str'" and "'<' not supported between instances of 'str' and 'int'" and sometimes messgaes did not match. 
I corrected error message in the function `nargsort` and tests, now all error messages match.

jbrockmendel commented on pandas/tests/base/test_misc.py:
```diff
@@ -151,6 +151,15 @@ def test_searchsorted(request, index_or_series_obj):
         mark = pytest.mark.xfail(reason="complex objects are not comparable")
         request.node.add_marker(mark)
+    elif any(isinstance(elem, int) for elem in obj.values[:]) and any(
+        isinstance(elem, str) for elem in obj.values[:]
```
this is a pretty awkward check.  maybe just check obj[0] and obj[1] with a comment the check is written with the mixed-int-string entry in mind?

natmokval replied:
agreed, I replaced the check with `tm.assert_mixed_int_string_entry`

jbrockmendel commented on pandas/tests/indexes/test_setops.py:
```diff
@@ -436,9 +464,20 @@ def test_intersect_unequal(self, index_flat, fname, sname, expected_name):
-        expected = index[1:].set_names(expected_name).sort_values()
-        tm.assert_index_equal(intersect, expected)
+        if (
+            len(index.values) > 0
```
maybe a tm.check_for_this_case to de-duplicate all these?  or just a new fixture that specifically excludes the relevant case?

natmokval replied:
thank you, I added `tm.assert_mixed_int_string_entry` and replaced the checks with it.

natmokval committed:
correct the error msg, add tm.assert_mixed_int_string_entry, correct tests

natmokval commented:
@jbrockmendel, I corrected my PR, could you please take a look at the update?

natmokval committed:
resolve conflict

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
fix tests

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

jbrockmendel commented on pandas/core/sorting.py:
```diff
@@ -436,7 +436,13 @@ def nargsort(
+    try:
+        indexer = non_nan_idx[non_nans.argsort(kind=kind)]
+    except TypeError as err:
+        msg = "'<' not supported between instances of 'int' and 'str'"
```
are there cases other than this particular test case that might hit this?

jbrockmendel commented on pandas/_testing/asserters.py:
```diff
@@ -1382,3 +1382,20 @@ def assert_metadata_equivalent(
             assert val == getattr(right, attr, None)
+
+
+def assert_mixed_int_string_entry(arr) -> bool:
```
maybe "is_mixed_..." instead of "assert_mixed_..."?

natmokval replied:
Thanks, I replaced `assert_mixed_int_string_entry` with `is_mixed_int_string_entry`.

jbrockmendel commented on pandas/_testing/asserters.py:
```diff
@@ -1382,3 +1382,20 @@ def assert_metadata_equivalent(
+
+def assert_mixed_int_string_entry(arr) -> bool:
+    """
+    Check that arr is mixed-int-string entry.
```
I think this will be clearer if it refers directly to the relevant fixture.

jbrockmendel replied:
i think this is an instance of a more general pain point in test writing:

```
def test_something(fixture):
    if the_fixture_value_is_this_particular_entry(fixture):
        ...
```

usually we either inspect the object like this assert_mixed_int_string_entry is doing or in some cases we pass `request` and do something hacky with request.fixturenames or something.  I wonder if there is a recommended pattern for like `pytest.get_active_fixture_id("index")` that would return "mixed-int-string" here

natmokval replied:
I coud't find a recommended pattern that would return `"mixed-int-string"` in this case. 
Instead I passed `request` and used `request.node.callspec.id` to return `"mixed-int-string"`.

Now I inspect the object by using `request.node.callspec.id` instead of  `is_mixed_int_string_entry`. The function `is_mixed_int_string_entry` is removed.
Sorry it took so long. Could you please take a look at this PR?

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
correct def nargsort, correect docs and rename assert_mixed_int_string_entry

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
add import _testing

natmokval committed:
fix pre-commit errors

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
fix pre-commit error

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval committed:
replace is_mixed_int_string_entry with request.node.callspec.id, correct tests

natmokval committed:
Merge branch 'main' into TST-mixed-int-string-Index

natmokval commented:
I added an entry: `"mixed-int-string"` in conftest's `indices_dict` and corrected tests, ci - green. @MarcoGorelli could you please take a look at this PR?

natmokval committed:
reolve conflicts

natmokval commented:
gentle ping @jbrockmendel

mroeschke commented:
Looks like this PR has gone stale and has conflicts now. Closing to clear the queue but if interested in contributing please resolve the conflicts and we can reopen
